posts/navarroaxel/17.git-rebase-interactive.md

```diff
@@ -0,0 +1,92 @@
+---
+title: Rewriting the history with Git rebase
+published: true
+description: How to squash commits and keep your branch up-to-date using git rebase 
+tags: 'git, bash, productivity, tooling'
+cover_image: ./assets/git-rebase.jpg
+id: 1949841
+---
+
+As when we work in a new feature or bug fix it's recommended to keep our branch up-to-date with the target branch (`main`, `development`, etc.), we can use the [`git rebase`](https://git-scm.com/docs/git-rebase) command instead of `git merge` in order to keep our Git history cleaner.
+
+The `git rebase` command reapplies our commits on top of another branch, rewriting the Git history; this is not recommended on shared branches (like `main`) because it causes divergence (and therefore conflict solving) and panic on your teammates 😬 since it is sometimes hard to resolve.
+
+## The basics
+
+To be clear about how `git rebase` works we're going to use the simplest example: you have 3 commits in your feature branch and you want to get the latest changes from `main`, so you can simply run the following command:
+
+```bash
+git rebase main
+```
+
+If you're lucky to get no conflicts then you'll have a new Git history with your 3 commits on the top of the `main` branch and you just need to override (`--force`) the remote branch with your new history.
+
+```bash
+git push --force origin feature/abc
+```
+
+But what if we have conflicts? Well, you're reapplying 3 commits so any of these could bring conflicts. Just resolve conflicts as always and then continue rebasing:
+
+```bash
+git add .
+git rebase --continue
+```
+
+Let's check a more dynamic scenario.
+
+## Squashing commits interactively
+
+If you made several commits while you prepared your work but you want to combine your commits into one before sending the pull request you can rebase against your target branch using the `--interactive` argument `git rebase --interactive main`, a `git-rebase-todo` file will be open in the default editor (`nano`, `vim`, etc.) where you can edit the commands that Git will execute to reapply your commits. Let's check this example:
```

Comment:
```suggestion
If you made several commits while you were preparing your work but now you want to combine your commits into one before creating the pull request then you can rebase against your target branch using the `--interactive` argument `git rebase --interactive main`. A `git-rebase-todo` file will be open in the default editor (`nano`, `vim`, etc.) where you can edit the commands that Git will execute to reapply your commits. Let's check this example:
```

Is it your target branch? Isn't it your current branch? If you use _target_ it sounds like you were talking about the remote